interpreter/luajit/luajit.go

```diff
@@ -0,0 +1,450 @@
+			// If this is a side trace, we need to add the spadjust of the root trace but
+			// only if they are different.
+			//https://github.com/openresty/luajit2/blob/7952882d/src/lj_gdbjit.c#L597
+			if t.root != 0 && traces[t.root].spadjust != t.spadjust {
+				spadjust += uint64(traces[t.root].spadjust)
+			}
+			p, err := l.addTrace(ebpf, fim, pr, t, uint64(g), spadjust)
+			if err != nil {
+				return err
+			}
+			newPrefixes = append(newPrefixes, p...)
+		}
+
+		logf("lj: new traces detected for pid(%v) added %d new traces with %d prefixes and removed %d prefixes",
+			pr.PID(), len(traces), len(newPrefixes), len(prefixes))
+
+		l.prefixesByG[g] = newPrefixes
+		l.traceHashes[g] = hash
+	}
+	return nil
+}
+
+func (l *luajitInstance) getGCproto(pt libpf.Address) (*proto, error) {
+	if pt == 0 {
+		return nil, nil
+	}
+	if gc, ok := l.protos[pt]; ok {
+		return gc, nil
+	}
+	gc, err := newProto(l.rm, pt)
+	if err != nil {
+		return nil, err
+	}
+	l.protos[pt] = gc
+	return gc, nil
+}
+
+// symbolizeFrame symbolizes the previous (up the stack)
+func (l *luajitInstance) symbolizeFrame(symbolReporter reporter.SymbolReporter,
+	funcName string, trace *libpf.Trace) error {
+	if l.previousFrame == nil || l.previousFrame.File == 0 {
+		return errors.New("previous frame not set")
+	}
+	ptAddr := libpf.Address(l.previousFrame.File)
+	pc := uint32(l.previousFrame.Lineno)
+	var line uint32
+	var fileName string
+	if ptAddr != C.LUAJIT_FFI_FUNC {
+		pt, err := l.getGCproto(ptAddr)
+		if err != nil {
+			return err
+		}
+		line = pt.getLine(pc)
+		fileName = pt.getName()
+	}
+	logf("lj: [%x] %v+%v at %v:%v", ptAddr, funcName, pc, fileName, line)
+	// The fnv hash Write() method calls cannot fail, so it's safe to ignore the errors.
+	h := fnv.New128a()
+	_, _ = h.Write([]byte(fileName)) //FIXME: needless allocation
+	_, _ = h.Write([]byte(funcName)) //FIXME: needless allocation
+	fileID, err := libpf.FileIDFromBytes(h.Sum(nil))
+	if err != nil {
+		return fmt.Errorf("failed to create a file ID: %v", err)
+	}
+	trace.AppendFrame(libpf.LuaJITFrame, fileID, libpf.AddressOrLineno(pc))
+	symbolReporter.FrameMetadata(fileID, libpf.AddressOrLineno(pc), util.SourceLineno(line),
+		pc, funcName, fileName)
+	return nil
+}
+
+// Symbolize is a little weird in lua since we need the caller frame to get the name of the
+// function being called. So we stash the info for the current frame and each symbolize call
+// actually symolizes the previous frame.  The unwinder emits a special frame with file 0 to
+// indicate the end of the lua stack.
+func (l *luajitInstance) Symbolize(symbolReporter reporter.SymbolReporter, frame *host.Frame,
+	trace *libpf.Trace) error {
+	if !frame.Type.IsInterpType(libpf.LuaJIT) {
+		return interpreter.ErrMismatchInterpreterType
+	}
+
+	// We need the calling frame to find the name of the current frame so save it and do it on next
+	// frame. This assumes that the harness calls Symbolize from top to bottom of the stack in
```

Comment:
We should probably document this assumption on the declaration of `Symbolize` in the `Instance` interface.

Reply:
:+1: